saiph/serializer.py

```diff
@@ -1,115 +1,120 @@
-            dtypes = valmap(str, obj.dtypes.to_dict())
-
-            # value_types is the type of a single cell in a column,
-            # which can be different from dtypes, especially when the dtype is 'object'
-            # or when we have nan-values together with integer values or stringified integers
-            value_types = {}
-            for col in obj.columns:
-                non_null_series = obj[col].dropna()
-                value_types[col] = (
-                    type(non_null_series.values[0]).__name__
-                    if not non_null_series.empty
-                    else "float"
-                )
-            data = obj.to_json(orient="index", default_handler=str)
-
-            # We cast to string and not to object, to transform the
-            # 'inferred_type' property (read-only) to string
-            # If we don't it remains at integer when decoding and using
-            # an index or column with `'0'`
-            index_type = (
-                "str" if str(obj.index.dtype) == "object" else str(obj.index.dtype)
-            )
-            columns_type = (
-                "str" if str(obj.columns.dtype) == "object" else str(obj.columns.dtype)
+    Parameters:
+        obj: object to encode
+    """
+    if isinstance(obj, np.ndarray):
+        data = obj.tolist()
+        return dict(__ndarray__=data, dtype=str(obj.dtype), shape=obj.shape)
+
+    if isinstance(obj, pd.Series):
+        data = obj.to_json(orient="index", default_handler=str)
+
+        # We cast to string and not to object, to transform the
+        # 'inferred_type' property (read-only) to string
+        # If we don't it remains at integer when using an index with `'0'`
+        index_type = "str" if str(obj.index.dtype) == "object" else str(obj.index.dtype)
+
+        return dict(__series__=data, dtype=str(obj.dtype), index_type=index_type)
+
+    if isinstance(obj, pd.DataFrame):
+        # orient='table' includes dtypes but doesn't work
+        dtypes = valmap(str, obj.dtypes.to_dict())
+
+        # value_types is the type of a single cell in a column,
+        # which can be different from dtypes, especially when the dtype is 'object'
+        # or when we have nan-values together with integer values or stringified integers
+        value_types = {}
+        for col in obj.columns:
+            non_null_series = obj[col].dropna()
+            value_types[col] = (
+                type(non_null_series.values[0]).__name__
+                if not non_null_series.empty
+                else "float"
             )
+        data = obj.to_json(orient="index", default_handler=str)
+
+        # We cast to string and not to object, to transform the
+        # 'inferred_type' property (read-only) to string
+        # If we don't it remains at integer when decoding and using
+        # an index or column with `'0'`
+        index_type = "str" if str(obj.index.dtype) == "object" else str(obj.index.dtype)
+        columns_type = (
```

Comment:
Looks like a duplicate of what's above. As there is a lot of logic to process in it, maybe it would be better to have a function doing that, with the comment in the docstring?

Reply:
It's not clear in the github diff, but I did not touch any of the code, I just got rid of the class and transformed it into a function.

`NumpyPandasEncoder.default` --> `numpy_pandas_json_encoding_hook`